File: priority.py
#!/usr/bin/env python3
"""Priority Scheduling - Non-preemptive (lower number = higher priority)"""

def priority_scheduling(processes):
    """processes: list of (pid, arrival_time, burst_time, priority)"""
    procs = sorted(processes, key=lambda p: p[1])  # sort by arrival
    time = 0
    result = []
    ready = []
    remaining = procs[:]
    
    while remaining or ready:
        # add newly arrived processes to ready queue
        while remaining and remaining[0][1] <= time:
            ready.append(remaining.pop(0))
        
        if ready:
            # pick highest priority (lowest number)
            ready.sort(key=lambda p: (p[3], p[1]))  # priority, then arrival
            pid, arrival, burst, priority = ready.pop(0)
            start = time
            completion = start + burst
            turnaround = completion - arrival
            waiting = start - arrival
            result.append((pid, arrival, burst, priority, start, completion, turnaround, waiting))
            time = completion
        else:
            # idle time
            if remaining:
                time = remaining[0][1]
    
    return result


def print_table(result):
    print("{:>4} {:>4} {:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format("PID", "AT", "BT", "PR", "ST", "CT", "TAT", "WT"))
    for row in result:
        print("{:>4} {:>4} {:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format(*row))
    n = len(result)
    avg_tat = sum(r[6] for r in result) / n
    avg_wt = sum(r[7] for r in result) / n
    print(f"\nAverage Turnaround Time = {avg_tat:.2f}")
    print(f"Average Waiting Time    = {avg_wt:.2f}")


if __name__ == '__main__':
    # (pid, arrival_time, burst_time, priority)
    processes = [
        (1, 0, 4, 2),
        (2, 1, 3, 1),
        (3, 2, 1, 3),
        (4, 3, 5, 2),
    ]
    res = priority_scheduling(processes)
    print("Priority Scheduling:\n")
    print_table(res)

File: priority.js
#!/usr/bin/env node
/**
 * Priority Scheduling - Non-preemptive (lower number = higher priority)
 */

function priorityScheduling(processes) {
    const procs = [...processes].sort((a, b) => a.arrival - b.arrival);
    let time = 0;
    const result = [];
    const ready = [];
    const remaining = [...procs];
    
    while (remaining.length > 0 || ready.length > 0) {
        // add newly arrived processes to ready queue
        while (remaining.length > 0 && remaining[0].arrival <= time) {
            ready.push(remaining.shift());
        }
        
        if (ready.length > 0) {
            // pick highest priority (lowest number)
            ready.sort((a, b) => a.priority - b.priority || a.arrival - b.arrival);
            const current = ready.shift();
            
            const start = time;
            const completion = start + current.burst;
            const turnaround = completion - current.arrival;
            const waiting = start - current.arrival;
            
            result.push({
                pid: current.pid,
                arrival: current.arrival,
                burst: current.burst,
                priority: current.priority,
                start,
                completion,
                turnaround,
                waiting
            });
            
            time = completion;
        } else {
            // idle time
            if (remaining.length > 0) {
                time = remaining[0].arrival;
            }
        }
    }
    
    return result;
}

function printTable(result) {
    console.log("PID".padStart(4), "AT".padStart(4), "BT".padStart(4), "PR".padStart(4),
                "ST".padStart(4), "CT".padStart(4), "TAT".padStart(5), "WT".padStart(5));
    
    for (const row of result) {
        console.log(
            String(row.pid).padStart(4),
            String(row.arrival).padStart(4),
            String(row.burst).padStart(4),
            String(row.priority).padStart(4),
            String(row.start).padStart(4),
            String(row.completion).padStart(4),
            String(row.turnaround).padStart(5),
            String(row.waiting).padStart(5)
        );
    }
    
    const n = result.length;
    const avgTAT = result.reduce((sum, r) => sum + r.turnaround, 0) / n;
    const avgWT = result.reduce((sum, r) => sum + r.waiting, 0) / n;
    
    console.log(`\nAverage Turnaround Time = ${avgTAT.toFixed(2)}`);
    console.log(`Average Waiting Time    = ${avgWT.toFixed(2)}`);
}

// Main execution
if (require.main === module) {
    const processes = [
        { pid: 1, arrival: 0, burst: 4, priority: 2 },
        { pid: 2, arrival: 1, burst: 3, priority: 1 },
        { pid: 3, arrival: 2, burst: 1, priority: 3 },
        { pid: 4, arrival: 3, burst: 5, priority: 2 }
    ];
    
    const result = priorityScheduling(processes);
    console.log("Priority Scheduling:\n");
    printTable(result);
}

module.exports = { priorityScheduling, printTable };
